CodeGenerator/Templates/Custom/s3/multipart_API.stencil

```diff
@@ -0,0 +1,250 @@
+                    return multipartDownloadPart(fileSize: contentLength, offset: 0)
+                } catch {
+                    return eventLoop.makeFailedFuture(error)
+                }
+        }
+        return result
+    }
+
+    /// Multipart download of a file from S3.
+    ///
+    /// - parameters:
+    ///     - input: The GetObjectRequest shape that contains the details of the object request.
+    ///     - partSize: Size of each part to be downloaded
+    ///     - filename: Filename to save download to
+    ///     - progress: Callback that returns the progress of the download. It is called after each part is downloaded with a value between 0.0 and 1.0 indicating how far the download is complete (1.0 meaning finished).
+    /// - returns: A future that will receive the complete file size once the multipart download has finished.
+    func multipartDownload(_ input: GetObjectRequest, partSize: Int = 5*1024*1024, filename: String, progress: @escaping (Double) throws -> () = { _ in }) -> Future<Int64> {
+        let eventLoop = AWSClient.eventGroup.next()
+        return eventLoop.submit {  () -> Foundation.FileHandle in
+            if FileManager.default.fileExists(atPath: filename) {
+                try FileManager.default.removeItem(atPath: filename)
+            }
+
+            FileManager.default.createFile(atPath: filename, contents: nil)
+
+            guard let fileHandle = FileHandle(forWritingAtPath: filename) else {
+                throw S3ErrorType.multipart.failedToOpen(file: filename)
+            }
+
+            return fileHandle
+        }.flatMap { (fileHandle) -> Future<Int64> in
+            var progressValue : Int64 = 0
+            let outputStream: (Data, Int64) throws -> () = {  data, fileSize in
+                fileHandle.write(data)
+                progressValue += Int64(data.count)
+                try progress(Double(progressValue) / Double(fileSize))
+            }
+            let download = self.multipartDownload(
+                input,
+                partSize: partSize,
+                on: eventLoop,
+                outputStream: outputStream
+            )
+            download.whenComplete { _ in
+                fileHandle.closeFile()
+            }
+            return download
+        }
+    }
+
+    /// Multipart upload of file to S3.
+    ///
+    /// - parameters:
+    ///     - input: The CreateMultipartUploadRequest structure that contains the details about the upload
+    ///     - inputStream: The function supplying the data parts to the multipartUpload. Each parts must be at least 5MB in size expect the last one which has no size limit.
+    /// - returns: A Future that will receive a CompleteMultipartUploadOutput once the multipart upload has finished.
+    func multipartUpload(_ input: CreateMultipartUploadRequest, inputStream: @escaping () throws -> Data?) -> Future<CompleteMultipartUploadOutput> {
+        var completedParts: [S3.CompletedPart] = []
```

Comment:
Obviously the FileHandle and eventloop changes you made for multipartDownload() will have to be implemented for the upload functions as well.